Grafana dashboard port

eljhkrr opened the pull request:
https://github.com/medic/cht-watchdog/issues/77

eljhkrr committed:
Feedback docs count

eljhkrr committed:
Feedback docs wip

eljhkrr committed:
Feedback docs widgets

eljhkrr committed:
Replication metrics

eljhkrr committed:
More replication metrics

eljhkrr committed:
Chrome version

eljhkrr committed:
SMS

eljhkrr committed:
Form completions

eljhkrr committed:
wip

eljhkrr committed:
Tasks

eljhkrr committed:
Form completions

eljhkrr committed:
Purge events

eljhkrr committed:
Last purge

mrjones-plip commented on watchdog-config/extra-sql-queries.yml:
```diff
@@ -75,3 +75,400 @@ users_with_insificient_disk:
+
+replication_by_status:
+  query: |
+    SELECT
```
this query returned 42k rows some of them with bad dates (`"2001-08-25"` ??) and old dates (`"2020-02-02"`) - should we limit this to maybe "last 90 days" instead?

eljhkrr replied:
Limited to last 90 days

mrjones-plip replied:
much better!  

though...did the table get truncated?  I'm only seeing ~900 rows without the 90 day limit - seems odd :thinking: 

But, let's resolve and move on!

mrjones-plip commented on watchdog-config/extra-sql-queries.yml:
```diff
@@ -75,3 +75,400 @@ users_with_insificient_disk:
+    - cht_instance:
+        usage: "LABEL"
+
+# replication_activity:
+#   query: |
+#     SELECT
+#         sms.partner_name,
+#         reported::varchar(255),
+#         delta_of_doc AS delta_docs,
+#         replace(monitoring_urls.url, 'https://', '') as cht_instance
+#     FROM app_monitoring_sms_error_and_users_docs_replication AS sms
+#     LEFT JOIN
+#         monitoring_urls ON sms.partner_name=monitoring_urls.partner_name
+#   metrics:
+#     - partner_name:
+#         usage: "LABEL"
+#     - reported:
+#         usage: "LABEL"
+#     - delta_docs:
+#         usage: "GAUGE"
+#     - cht_instance:
+#         usage: "LABEL"
```
no need to store this here!  

```suggestion
```

eljhkrr replied:
Sorry I didn't add a comment earlier on this. The query was causing 500 errors within SQL exporter and I couldn't figure out why. It's required for the docs synced per day metric and runs well on the PG client. No useful pointers in container logs. I could use your help with diagnosing why the 500s are happening.

mrjones-plip replied:
hmmm - hard for me to say without being able to reproduce it.   I did play around with the query and see that as is it's returning 21k results and this only returns ~800 which seems more sane. Maybe helpful as a starting point to debug by reducing data it's processing?

```sql
SELECT
	sms.partner_name,
	reported::date,
	delta_of_doc AS delta_docs,
	replace(monitoring_urls.url, 'https://', '') as cht_instance
FROM 
	app_monitoring_sms_error_and_users_docs_replication AS sms
LEFT JOIN
  monitoring_urls ON sms.partner_name = monitoring_urls.partner_name
WHERE 
	monitoring_urls.url is not NULL and
	delta_of_doc > 0 and
	reported > current_date - interval '30' day;
```

eljhkrr replied:
Reducing data processing worked, thank you

mrjones-plip replied:
new query looks great and is fast!

mrjones-plip commented on watchdog-config/extra-sql-queries.yml:
```diff
@@ -75,3 +75,400 @@ users_with_insificient_disk:
+
+initial_replications:
+  query: |
+    SELECT
```
is the plan to chart these over time?  I see, for example `bhojpur-ne`  returns 61 results and only 55 with   `count_initial_replications` values of  not `0` 

just making sure this is helpful data to query!

eljhkrr replied:
Great suggestion limited to results with `count_initial_replications > 0`

mrjones-plip replied:
great! I see just about every query has this now - nice work.

mrjones-plip commented on watchdog-config/extra-sql-queries.yml:
```diff
@@ -75,3 +75,400 @@ users_with_insificient_disk:
+
+tasks_by_user:
+  query: |
+    SELECT
```
this query returns 32k results - any way to trim it down?

eljhkrr replied:
Limited to last 90 days

mrjones-plip replied:
please revisit this - it looks like it didn't get added to [this query when I check the most recent version](https://github.com/medic/cht-app-monitoring-data-ingest/blob/34e28d277794ec2607bd0d4cc6178b9ad5a2f5a6/watchdog-config/extra-sql-queries.yml#L405).

eljhkrr replied:
Updated - I had limited tasks_by_state to last 90 days and realized tasks_by_user wouldn't have a reduction in row count by limiting time window. However, it does makes sense to only view tasks in the last 90 days per user.

mrjones-plip replied:
while you did add an extra where clause `AND day >= NOW() - INTERVAL '90 days'` , there is no `day` column so the query fails.

Maybe we [can try to do a](https://stackoverflow.com/questions/2318398/how-can-i-limit-number-of-results-by-a-specific-column-in-postgresql) `PARTITION BY`?  I just found that searching around, I don't have a good option here :shrug:

mrjones-plip commented on watchdog-config/extra-sql-queries.yml:
```diff
@@ -75,3 +75,400 @@ users_with_insificient_disk:
+
+form_completions:
+  query: |
+    SELECT
```
same as above - this one returns 18k results which seems a bit verbose

eljhkrr replied:
Limited to last 90 days

mrjones-plip replied:
please revisit this - it looks like it didn't get added to [this query when I check the most recent version](https://github.com/medic/cht-app-monitoring-data-ingest/blob/34e28d277794ec2607bd0d4cc6178b9ad5a2f5a6/watchdog-config/extra-sql-queries.yml#L405).

eljhkrr replied:
Updated

eljhkrr replied:
Walking this back as limitng by time frame will not have a significant impact on the result set size. It has to aggregate by all unique form ID's & users.

eljhkrr replied:
Will limit time window in the upstream function

mrjones-plip replied:
I defer to you, but I still can't see how MoH Mali CHW, for example, would actually want to look at almost 2.3k rows in anything but a much more summarized way.  If you thinks it's still useful to get this may rows back (18k total) we can keep it!

mrjones-plip commented on watchdog-config/extra-sql-queries.yml:
```diff
@@ -75,3 +75,400 @@ users_with_insificient_disk:
+
+purge_logs:
+  query: |
+    SELECT
```
this one returns 17k results - just about every single one (or all of them!?!) are for `partner_name` of `itech_aurum_scaleup` - maybe exclude them? Or purge that table and re-populate?

eljhkrr replied:
Using partition by to limit results to top 5 by partner

mrjones-plip replied:
ah - this is just right! I see we have an inner `SELECT` that returns 17k rows, but the outer `SELECT` culls that down to 5 per partner, just as you said.  Having this expensive query run only every 12 hours and return a small amount of important results is great!

mrjones-plip reviewed (changes requested):
yay!  great work here on getting a bunch of queries in .

The JSON is too big to review, manually, so I dove into the SQL.  why don't you tidy these up per feedback and resolve the merge conflicts.  then I'll spin it up in my dev instance and take a look at the new dashboard's JSON by reviewing the rendered view in grafana.

See review comments and also: I think we can improve the performance by always excluding results that don't have a partner.  We can append this to every query:

```sql
      where monitoring_urls.url is not NULL
```

On the `invalid_doc_errors` query for example, this went from ~8k results to ~5k results!

eljhkrr committed:
wip

eljhkrr committed:
Merge branch 'main' into grafana

eljhkrr committed:
Filter empty URLs

eljhkrr commented:
I've resolved conflicts to unblock visual testing of the dashboard. Some of the performance improvement suggestions need data inspection and I'll get back on them soon.

mrjones-plip commented on watchdog-config/extra-sql-prometheus.yml:
```diff
@@ -1,5 +1,6 @@
 scrape_configs:
   - job_name: 'extra-sql'
-    scrape_interval: 12h
+    scrape_interval: 300s
+    scrape_timeout: 180s
```
While fine to keep at `300s` for testing, please don't commit this value!  `12h` is more than enough given the data!  Also `12h` will greatly reduce the load on the RDBMS server

```suggestion
    scrape_interval: 12h
    scrape_timeout: 180s
```

eljhkrr replied:
Updated

mrjones-plip commented:
Working to show the new queries in my local dev environment and noticed that the SQL Exporter is returning over 50MB of raw text.  This is due to the largely unbound queries returning way more than we need. We should work to reduce the data in each query by 10-100x:

![image](https://github.com/medic/cht-app-monitoring-data-ingest/assets/8253488/9d1357cb-4e5f-43ec-8772-e6907335bd7b)

mrjones-plip commented:
@eljhkrr - Wow wow!!! You've done a really impressive amount of work on this PR - nice job!  I got it to load in my local dev instance and it's a beauty to behold!! Here it is:

![image](https://github.com/medic/cht-app-monitoring-data-ingest/assets/8253488/f05eeded-f27c-4cb4-83d1-daba1a3cf65c)

That said, we should wait to merge and fix up some items:
* Greatly reduce query size which in turn will reduce data returned to UI which in turn will make UI much more responsive. See above comment for [an example](https://github.com/medic/cht-app-monitoring-data-ingest/pull/108#issuecomment-2050739677).
* "Purging events" for me has all the dates mashed together on the X axis - I can't read any of them
* too many events to be practical in manual and automatic feedback docs (1000s in an 6 line table) 
* can't quite tell if it's due to my setup, but these panels showed "No data" when testing with `muso-mali.app.medicmobile.org`:
   * days since succ. purge
   * users with ins. chrome ver
   * users with ins. disk
   * docs replicated per day
* gateway state: I think this should just be removed? it shows loong values (see end of this bullet) and also shows "state" of `0` which doesn't make much sense? `{__name__="gateway_health_due", cht_instance="muso-mali.app.medicmobile.org", delivered="0", gateway_state="Gateway Healthy", instance="extra_sql_exporter:9187", job="extra-sql", partner_name="musoapp", reported="2023-12-13", scheduled="0", server="172.17.0.1:5431"}` 
* "Days since last sync" and "Feedback docs by type"  these panels were showing time series but with no real data?
* "form  completions by user" - is there a way we can have a drop down to select a user?  also - can we format percent so `1` shows as `100%` and `0.857` shows as `85.7%` ?
* "form completions" - is there a way we can have a drop down to select a form?  also - can we format percent so `1` shows as `100%` and `0.857` shows as `85.7%` ?

eljhkrr committed:
Limit result size

eljhkrr committed:
Scrape interval

mrjones-plip reviewed (changes requested):
Looking good! 

I still see some cleanup to do:
* `manual_feedback_docs` - 2k result , mostly for brac and vhtapp - not limited to 90 days
* `automatic_feedback_docs` - is ~1k results, maybe OK, but still not sure how viewing 100s of results in a table is helpful - still want to keep?
*  `replication_activity` - 3.1k results - and is it helpful to return restuls of `delta_docs` of `0`? I see adding `and delta_of_doc > 0` reduces results to 2.6k
* `chw_engagement` - returns `0` results is that correct?
* `gateway_health` - limit to 90 days, yeah? 3.8k results going back to early Dec 2023
* `invalid_doc_errors` and `tasks_by_state` returns 4.9k and 5.6k results respectively - is this helpful to have this fine granularity?  might be ok!

eljhkrr commented:
Thanks @mrjones-plip for the feedback, I'll review and get back

eljhkrr commented:
> Greatly reduce query size which in turn will reduce data returned to UI which in turn will make UI much more responsive. See above comment for https://github.com/medic/cht-app-monitoring-data-ingest/pull/108#issuecomment-2050739677.

A number of metrics have been limited to the last 90 days. This will be a continuous effort as we use the dashboard.

> "Purging events" for me has all the dates mashed together on the X axis - I can't read any of them

Updated to improve data point display on x-axis

> too many events to be practical in manual and automatic feedback docs (1000s in an 6 line table)

Limited to 50 rows in a paginated view

> can't quite tell if it's due to my setup, but these panels showed "No data" when testing with muso-mali.app.medicmobile.org:
> days since succ. purge
> users with ins. chrome ver
> users with ins. disk
> docs replicated per day

The backing queries are extracting data well. It's possible that there are no records for Muso with these data points. To be 100% sure, this necessitates diving into project data.

> gateway state: I think this should just be removed? it shows loong values (see end of this bullet) and also shows "state" of 0 which doesn't make much sense? {__name__="gateway_health_due", cht_instance="muso-mali.app.medicmobile.org", delivered="0", gateway_state="Gateway Healthy", instance="extra_sql_exporter:9187", job="extra-sql", partner_name="musoapp", reported="2023-12-13", scheduled="0", server="172.17.0.1:5431"}

Fixed presentation of gateway state. Negative values need to be investigated:
![image](https://github.com/medic/cht-app-monitoring-data-ingest/assets/4529018/05d62ab6-6f7d-4023-a22a-bd7dbe6088b1)

> "Days since last sync" and "Feedback docs by type" these panels were showing time series but with no real data?

app_monitoring_chw_engagement is the backing data source for 'days since last sync' and doesn't have any data

'Feedback docs by type' fixed
![image](https://github.com/medic/cht-app-monitoring-data-ingest/assets/4529018/144b85a8-fe09-444f-be27-af24d5ee3378)


> "form completions by user" - is there a way we can have a drop down to select a user? also - can we format percent so 1 shows as 100% and 0.857 shows as 85.7% ?
> "form completions" - is there a way we can have a drop down to select a form? also - can we format percent so 1 shows as 100% and 0.857 shows as 85.7% ?

Added column filters and % formatting

eljhkrr committed:
Iteration

mrjones-plip commented:
awesome!  Thanks for working though all my requests for improvements.  I'll do another round of review with the optimized queries and grafana front end in the coming day or two.

eljhkrr commented:
Thank you, I appreciate your insights! I have one last set of feedback comments to get through hopefully before your next review.

mrjones-plip commented:
ah - good good!  I'll keep an eye out and do my next review after I see another round updates here.

eljhkrr committed:
Feedback iteration

eljhkrr commented:
> manual_feedback_docs - 2k result , mostly for brac and vhtapp - not limited to 90 days
> automatic_feedback_docs - is ~1k results, maybe OK, but still not sure how viewing 100s of results in a table is helpful - still want to keep?

Updated to use a single data source for feedback documents. Table limited to last 50 results with column filters for search.

> replication_activity - 3.1k results - and is it helpful to return restuls of delta_docs of 0? I see adding and delta_of_doc > 0 reduces results to 2.6k

Limited to rows where delta_docs > 0

> chw_engagement - returns 0 results is that correct?

chw_engagement returns 0 rows for the last full month as configured. Reopening #14 to consider alternative interpretation.

> gateway_health - limit to 90 days, yeah? 3.8k results going back to early Dec 2023

Limited to last 90 days

> invalid_doc_errors and tasks_by_state returns 4.9k and 5.6k results respectively - is this helpful to have this fine granularity? might be ok!

Yes it's not possible to simplify further according to chart needs

mrjones-plip commented on watchdog-config/extra-sql-queries.yml:
```diff
@@ -53,58 +79,373 @@ dwh_users_with_insificient_disk:
+    FROM app_monitoring_sms_error_and_users_docs_replication AS sms
+    LEFT JOIN
+        monitoring_urls ON sms.partner_name = monitoring_urls.partner_name
+    WHERE monitoring_urls.url IS NOT NULL AND reported::date >= NOW() - INTERVAL '90 days' AND delta_of_doc > 0
```
Minor point, but this, and all other `WHERE` clauses would be a bit easier to read with line breaks (this line is especially wide):

```suggestion
    WHERE 
        monitoring_urls.url IS NOT NULL AND 
        reported::date >= NOW() - INTERVAL '90 days' AND 
        delta_of_doc > 0
```

mrjones-plip reviewed (changes requested):
Nice work here!  I think we're a lot closer.

I still have big concerns about the huge size of the exported SQL metrics. While it [was 56MB uncompressed](https://github.com/medic/cht-app-monitoring-data-ingest/pull/108/files#issuecomment-2050739677) and it's been reduced to 32MB, this still seems like a tone of data.  It's 180k lines of text

```
➜  cht-watchdog git:(81-sql-exporter) ✗  curl -O http://172.30.0.3:9187/metrics
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 32.8M    0 32.8M    0     0  1162k      0 --:--:--  0:00:28 --:--:-- 8987k

➜  cht-watchdog git:(main) ✗ wc -l metrics
180077 metrics
```

Conversely, I check and the in-dashboard experience appears to load much better due to our optimizations.  For example, the [prior 11mb of JSON](https://github.com/medic/cht-app-monitoring-data-ingest/pull/108/files#issuecomment-2050739677) is now down to a tame 198k:

![image](https://github.com/medic/cht-app-monitoring-data-ingest/assets/8253488/39cda9d6-f115-4f9e-8286-1c2946496f0c)

I'll bring a discussion up in `#product-allies` channel to see what some other folks think of the sizes we're seeing here as far as what's reasonable.  As such, I'm flagging this as "Request Changes" and we can decide how to proceed together.

I otherwise made a few other feedback points.

mrjones-plip commented on watchdog-config/extra-sql-dashboard.json:
```diff
@@ -367,224 +322,2743 @@
+          "useBackend": false
+        }
+      ],
+      "title": "# docs replicated per day",
```
I'm seeing a label that's not readable b/c it needs to be culled down:

![image](https://github.com/medic/cht-app-monitoring-data-ingest/assets/8253488/0c598019-9644-4302-be83-2c961709d901)

should probably just hide the labels?  this isn't helpful:

```
{__name__="replication_activity_delta_docs", cht_instance="disc-mali.ml", instance="extra_sql_exporter:9187", job="extra-sql", partner_name="moh_mali_chw", reported="2024-02-05", server="172.17.0.1:5431"}
```

Also...is this data right?  There were exactly 2,500 docs replicated every time?  this seems either inaccurate or not helpful maybe?

mrjones-plip commented on watchdog-config/extra-sql-dashboard.json:
```diff
@@ -367,224 +322,2743 @@
+          "useBackend": false
+        }
+      ],
+      "title": "Feedback docs by type",
```
we should clean up the labels on this:

![image](https://github.com/medic/cht-app-monitoring-data-ingest/assets/8253488/c1e76919-71cd-4dd6-825d-99fcd779824f)

I think this:

```
feedback_docs_source_count {__name__="feedback_docs_source_count", cht_instance="disc-mali.ml", instance="extra_sql_exporter:9187", job="extra-sql", partner_name="moh_mali_chw", server="172.17.0.1:5431", source="manual"}
```

should just be `manual` I think?

mrjones-plip commented:
slack thread started here.  tl;dr:

> All else being equal, I propose we push this live and see if it breaks.  We run the scrape every [12hrs](https://github.com/medic/cht-app-monitoring-data-ingest/blob/b923ed21aea22220dba3d510fca96887cf6d3512/watchdog-config/extra-sql-prometheus.yml#L3) so it won't impact RDBMS much so should be safe.  Let me know if you disagree!

mrjones-plip reviewed (approved):
With https://github.com/medic/medic-infrastructure/issues/961 being fixed, [current extra sql queries](https://watchdog.app.medicmobile.org/d/a71db640-cc40-452c-aa92-222a9b49d43b/cht-admin-extra-sql?orgId=1&refresh=5s&var-cht_instance=disc-mali.ml) are now working again :sweat_smile: . 

Let's merge this, get it live and we can iterate to fix things that are broken as well as migrate to the new sql exporter per https://github.com/medic/cht-watchdog/issues/81 which is live and field tested.

eljhkrr commented:
Agreed, I'll pull pending items into a separate issue
